metaflow/subprocess_manager.py

```diff
@@ -0,0 +1,330 @@
+
+        return self.commands.get(pid, None)
+
+    def cleanup(self) -> None:
+        """Clean up log files for all running subprocesses."""
+
+        for v in self.commands.values():
+            v.cleanup()
+
+
+class CommandManager(object):
+    """A manager for an individual subprocess."""
+
+    def __init__(
+        self,
+        command: List[str],
+        env: Optional[Dict[str, str]] = None,
+        cwd: Optional[str] = None,
+    ):
+        self.command = command
+
+        self.env = env if env is not None else os.environ.copy()
+        self.cwd = cwd if cwd is not None else os.getcwd()
+
+        self.process = None
+        self.run_called: bool = False
+        self.log_files: Dict[str, str] = {}
+
+        signal.signal(signal.SIGINT, self.handle_sigint)
+
+    async def __aenter__(self) -> "CommandManager":
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        self.cleanup()
+
+    def handle_sigint(self, signum, frame):
+        """Handle the SIGINT signal."""
+
+        print("SIGINT received.")
+        asyncio.create_task(self.kill())
+
+    async def wait(
+        self, timeout: Optional[float] = None, stream: Optional[str] = None
+    ) -> None:
+        """Wait for the subprocess to finish, optionally with a timeout and optionally streaming its output."""
+
+        if timeout is None:
+            if stream is None:
+                await self.process.wait()
+            else:
+                await self.emit_logs(stream)
+        else:
+            try:
+                if stream is None:
+                    await asyncio.wait_for(self.process.wait(), timeout)
+                else:
+                    await asyncio.wait_for(self.emit_logs(stream), timeout)
+            except asyncio.TimeoutError:
+                command_string = " ".join(self.command)
+                await self.kill()
+                print(
+                    "Timeout: The process: %s with command: '%s' didn't complete within %s seconds."
+                    % (self.process.pid, command_string, timeout)
+                )
+
+    async def run(self):
+        """Run the subprocess, streaming the logs to temporary files"""
+
+        self.temp_dir = tempfile.mkdtemp()
+        stdout_logfile = os.path.join(self.temp_dir, "stdout.log")
+        stderr_logfile = os.path.join(self.temp_dir, "stderr.log")
+
+        try:
+            # returns when process has been started,
+            # not when it is finished...
+            self.process = await asyncio.create_subprocess_exec(
+                *self.command,
+                cwd=self.cwd,
+                env=self.env,
+                stdout=open(stdout_logfile, "w"),
+                stderr=open(stderr_logfile, "w"),
+            )
+
+            self.log_files["stdout"] = stdout_logfile
+            self.log_files["stderr"] = stderr_logfile
+
+            self.run_called = True
+            return self.process
```

Comment:
we seem to only use the PID where we call this. Can we just return that?

Reply:
done